### PR TITLE
Import you tube playlist

### PR DESCRIPTION
**Pull request by muydanny**
Import playlist from youtube spec

View it here now a link

Associates each video stored in the DB is associated with a tutorial

Tutorial video count > 50 spec

Can import a playlist with more than 50 videos + edge case of larger playlists

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -14,32 +14,35 @@
...
       <% @repos.each do |repo| %>
         <ul>
           <li><a href= "<%= repo.url %>"><%= repo.name %></li></a>
         </ul>
       <% end %>
-      <section class="github_followers">
-        <h1><%= @followers.length %> Followers</h1>
-        <% @followers.each do |follower| %>
+    <% end %>
+
```
Comment (muydanny) [inline app/views/users/show.html.erb:24]: I was getting syntax errors so I spaced things out for better readability (spelling?)

File: app/views/admin/tutorials/new.html.erb [hunk 2]
```diff
@@ -15,4 +16,13 @@
...
 
   <%= f.submit "Save", class: "mt2 btn btn-primary mb1 bg-teal"  %>
+
+  <h1>Add New Youtube Playlist</h1>
+  <section class="import-youtube">
+    <%= f.fields_for :playlist do |builder| %>
+      <%= f.label 'Playlist ID' %>
+      <%= f.text_field :playlist_id, class: "block col-4 field" %>
+    <% end %>  
+    <%= link_to 'Import Youtube Playlist', "#", "data-action" => "click->tutorials#updateOrder", class: "video-position-submit btn btn-outline", id: @tutorial.id %>
```
Comment (muydanny) [inline app/views/admin/tutorials/new.html.erb]: Looked through other files and copied that set up for this code here. This also ensures the css is the same for the new pages we make

File: spec/features/admin/new_tutorial_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    save_and_open_page
+    expect(current_path).to eq('/admin/dashboard')
+    expect(page).to have_content('Successfully created tutorial. View it here.')
+    
+    click_link('View it here.')
+    id = Tutorial.last.id
+    expect(current_path).to eq("/tutorials/#{id}")
+
+  end
+end
```
Comment (muydanny) [inline spec/features/admin/new_tutorial_spec.rb:69]: Ill add to the project board that we delete comment out code.

File: app/controllers/admin/tutorials_controller.rb [hunk 4]
```diff
@@ -6,9 +6,9 @@ def edit
   def create 
     tutorial = Tutorial.create(new_tutorial_params)
     if tutorial.save
-      YoutubeResults.new.create_videos(tutorial)
-      flash[:notice] = "Successfully created tutorial. View it here."
+      YoutubeService.new.create_videos(tutorial)
       redirect_to admin_dashboard_path
+      flash[:notice] = "Successfully created tutorial. #{view_context.link_to("View it here.", tutorial_path(tutorial.id))}"
```
Comment (muydanny) [inline app/controllers/admin/tutorials_controller.rb:13]: view_context is a cool rails method I came across. This allows you to have a link within your flash message.

File: app/services/youtube_service.rb [hunk 5]
```diff
@@ -1,4 +1,10 @@
 class YoutubeService
+
+  def create_videos(tutorial)
```
Comment (muydanny) [inline app/services/youtube_service.rb]: In a previous commit I had planned on making YoutubeResults but found this YoutubeService.

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -16,7 +16,7 @@
     <main class="px4">
       <div class="flash-message">
         <% flash.each do |name, msg| %>
-          <%= content_tag :div, msg, class: name %>
+          <%= content_tag :div, msg.html_safe, class: name %>
```
Comment (muydanny) [inline app/views/layouts/application.html.erb:19]: This has to do with view_context. According to the documentation I came I across we have to make it html_safe. Not 100 percent sure why but the link doesnt work without it. From what I gathered it has something to do with security.

File: app/services/youtube_service.rb [hunk 7]
```diff
@@ -2,7 +2,20 @@ class YoutubeService
 
   def create_videos(tutorial)
     url = "/youtube/v3/playlistItems?playlistId=#{tutorial.playlist_id}"
-    res = get_json(url, {})  
+    res = get_json(url, {part: "snippet"})  
+  end
+  
+  # could go into a youtube poro
+  def playlist_video_params(tutorial)
```
Comment (muydanny) [inline app/services/youtube_service.rb:23]: There are probably a couple things that could be renamed looking at this again.

File: app/services/youtube_service.rb [hunk 8]
```diff
@@ -1,14 +1,28 @@
...
+    if nextPageToken
+      params[:pageToken] = nextPageToken
+    end
+
+    res = get_json(url, params)    
+    videos << res[:items]
+    
+    if res[:nextPageToken]
+      # using recursion 
+      create_videos(tutorial, videos, res[:nextPageToken])
```
Comment (muydanny) [inline app/services/youtube_service.rb:16]: This part is cool. Im using recursion which is pretty much calling a method which has the same method within it. Kinda like a loop but it is only called it a nextPageToken is passed through. So this allows us to display more that 50 videos.

File: app/controllers/admin/tutorials_controller.rb (hunk 4 above)
Comment (jhgould) [reply app/controllers/admin/tutorials_controller.rb:13]: Wait thats awesome. Im definitely going to use that in the future.

**Pull request merged**